torch/csrc/autograd/input_buffer.cpp

```diff
@@ -268,19 +288,30 @@ void InputBuffer::add(
       auto event = c10::Event{device_type};
       event.record(*opt_producer_stream);
       accum_stream->wait(event);
-      record_stream_any_impl(var, *accum_stream);
+      record_stream_used = true;
+      if (!autograd_avoid_record_streams()) {
+        record_stream_any_impl(var, *accum_stream);
+      }
     }
     if (*accum_stream != *ready_stream) {
       accum_stream->wait(*ready_event);
       // This is redundant for case A, but needed for case C
-      record_stream_any_impl(buffer[pos], *accum_stream);
+      record_stream_used = true;
+      if (!autograd_avoid_record_streams()) {
+        record_stream_any_impl(buffer[pos], *accum_stream);
+      }
     }
     // 2)
     c10::OptionalStreamGuard stream_guard{accum_stream};
     accumulate(buffer, pos, std::move(var));
     // 3)
     auto event = c10::Event{device_type};
     event.record(*accum_stream);
+    if (autograd_avoid_record_streams() && record_stream_used) {
+      auto event_for_record_stream = c10::Event{device_type};
+      event_for_record_stream.record(*accum_stream);
+      opt_producer_stream->wait(event_for_record_stream);
+    }
```

Comment:
Not sure this is sufficient because both producers (possibly different streams) should wait for the accumulate stream. (I think this only matters for case C?)

We may also want to do this before accumulating because var (which is moved into that function) will lose all references before we return from that.

Reply:
> Can we stash a reference before accumulating instead? I'm thinking that it would be unsafe to wait for the event before its last use (the accumulation)?

Good point, that sounds better.